File: Employee/controller.py
from os import access
from pyexpat.errors import messages
from urllib import response
from wsgiref.validate import validator
from Employee.validation import validation_check, validation_insert, validation_insert_id, validation_update
from Utilities import db
from flask import Flask,jsonify,request,Blueprint
from timeit import default_timer as timer
from datetime import datetime,timedelta
from Utilities import db
import hashlib
from flask_jwt_extended import create_access_token,decode_token
from flask_jwt_extended import get_jwt_identity
from flask_jwt_extended import jwt_required
from flask_jwt_extended import JWTManager


def give_response(data,message,start_time):
    end_time = datetime.now()
    duration = end_time - start_time
    response = {
        "start_time": start_time.strftime("%H:%M:%S.%f"),
        "success": True,
        "data": data,
        "message":message,
        "end_time": end_time.strftime("%H:%M:%S.%f"),
        "duration":duration.total_seconds()
    }
    return response

def give_response_access_token(data,message,access_token,start_time):
    end_time = datetime.now()
    duration = end_time - start_time
    response = {
        "start_time": start_time.strftime("%H:%M:%S.%f"),
        "success": True,
        "data": data,
        "message":message,
        "end_time": end_time.strftime("%H:%M:%S.%f"),
        "duration":duration.total_seconds(),
        "access_token":access_token
    }
    return response

def test_authorization():
    auth = str(request.headers['Authorization']).split(' ')[1]
    output= decode_token(auth)['sub']
    return output

def give_hash(input):
    hash = hashlib.md5(input.encode())
    return hash.hexdigest()

def check_employee():
    email = test_authorization()
    query = "select email from employee where email = '%s';"%(input[email])
    output = db.read_from_table(query)
    if output[0][0]==1:
        return True
    else:
        return "user not found"


def read_employee():
        start_time = datetime.now()
        query = "select * from employee;"
        var = db.get_all(query)
        return jsonify(give_response(data=var, message='operation successful', start_time=start_time))

def read_employee_id():
        start_time = datetime.now()
        auth = str(request.headers['Authorization']).split(' ')[1]
        output= decode_token(auth)['sub']
        if (output ==1):
            value = request.args
            v= validation_insert_id(value)
            if isinstance(v,bool):
                query = "select * from employee where id = %s;"%(int(value["id"]))
                var = db.get_all_id(query)
                return jsonify(give_response(data=var, message='operation successful', start_time=start_time))
            else:
                return jsonify({ "message" : v })
        else:
            return jsonify(give_response(data=[], message='operation falied', start_time=start_time))


def insert_employee():
    start_time = datetime.now()
    if isinstance(check_employee(),bool):
        input = {}
        input = request.get_json()
        v= validation_insert(input)
        if isinstance(v,bool):
            query = "INSERT INTO employee (id, first_name, last_name, email, gender, phone ,password) VALUES (%s, '%s', '%s', '%s', '%s', '%s', '%s');"%(input["id"],input["first_name"],input["last_name"],input["email"],input["gender"],input["phone"],give_hash(input["password"]))
            messages = db.insert(query)
            var = jsonify(give_response(data=[messages], message="row is inserted", start_time=start_time))  
            return var 
        else:
            return jsonify({ "message" : v })
    else:
         jsonify(give_response(data=[], message="operation failed", start_time=start_time))  


def update_employee():
    start_time = datetime.now()
    if isinstance(check_employee(),bool):
        input = {}
        input = request.get_json()
        v= validation_update(input)
        if isinstance(v,bool):
            query = "UPDATE employee SET first_name = '%s' WHERE id = %s;"%(input["first_name"],input["id"])
            messages = db.update(query)
            return jsonify(give_response(data=messages, message="row is updated", start_time=start_time))
        else:
            return jsonify({ "message" : v })
    else:
        return jsonify(give_response(data=[], message="operation falied", start_time=start_time))


def delete_employee():
    start_time = datetime.now()
    if isinstance(check_employee(),bool):
        value = request.args
        v= validation_insert_id(value)
        if isinstance(v,bool):
            query = "DELETE FROM employee WHERE id = %s;"%(int(value["id"]))
            messages = db.delete(query)
            return jsonify(give_response(data=messages, message="Deleted the data", start_time=start_time))
        else:
            return jsonify({ "message" : v })
    else:
            return jsonify(give_response(data=[], message="operation falied", start_time=start_time))


def truncate_employee():
    start_time = datetime.now()
    query = "truncate table employee;"
    message = db.truncate(query)
    return jsonify(give_response(data=[], message=message, start_time=start_time))

def student_login():
    start_time = datetime.now()
    input = {}
    input = request.get_json()
    v= validation_check(input)
    if isinstance(v,bool):
            query = f"select password from employee where email = '{input['email']}'"
            message = db.get_all(query)
            if give_hash(input['password']) == message[0][0]:
                print("HELLO")
                access_token = create_access_token(identity=input['email'])
                return jsonify(give_response_access_token(data=[], message="login successful",access_token=access_token, start_time=start_time))
            else:
                return jsonify(give_response(data=[], message="wrong username or password", start_time=start_time))
    else:
        return jsonify({ "message" : v })

File: Employee/controller.js
const executor = require("../Utilities/db");
const config = require("../Config/config");
const errorHandle = require('../Utilities/error_handler')
const responsehandler = require("../Utilities/response_handler");
const md5 = require("md5");
const jwt = require('jsonwebtoken')

const readEmployee = async (req, res,next) => {
    try {
        console.log("READING")
        let result = await executor
            .executeQuery("select * from employee")
            .catch(function reject(error) {
                throw error;
            });
            console.log("------------------------------------------------------------")
            console.log(result)
        //return res.status(200).send(result)    
        return res.status(200).json(responsehandler.makeResponse("query successful!!", result));
    // } catch (error) {
    //     return res.status(404).json(responsehandler.makeErrorResponse("error while reading employee data", error.message));
    }
    catch(err){
        let errorInstance = errorHandle.errorHandle(500,"Internal server error",err)
        next(errorInstance)
    }
};


const readEmployeeid = async (req, res,next) => {
    try {
        let email = req.email;
        // console.log("----------------------------------------")
        // console.log(email);
        const inputData = [email];
        query = `select * from employee where email = ?`;
        result = await executor
            .executeQuery(query, inputData)
            .catch(function reject(error) {
                throw error;
            });
            return res.status(200).send(result)
        }
    catch(err){
        let errorInstance = errorHandle.errorHandle(500,"Internal server error",err)
        next(errorInstance)
    }
};




const signupEmployee = async (req, res) => {
    try {
        let id = req.body.id;
        let first_name = req.body.first_name;
        let last_name = req.body.last_name;
        let email = req.body.email;
        let gender = req.body.gender;
        let phone = req.body.phone;
        let password = req.body.password;
        let passwordDigit = md5(password);
        const inputData = [[[id, first_name, last_name, email, gender, phone, passwordDigit]]];
        query = `insert into employee(id, first_name, last_name, email, gender, phone, password) values ?`;
        result = await executor
            .executeQuery(query, inputData)
            .catch(function reject(error) {
                throw error;
            });
        const tokens = jwt.sign(
            { email },
            config.token,
            {
                expiresIn: "5s",
            }
        );
        // save user token
        return res.status(200).json(responsehandler.makeTokenResponse("signup sucessfull", tokens));

    } catch (error) {
        return res.status(500).json(responsehandler.makeErrorResponse("error while inserting employee data", error.message));
    }
};



const insertEmployee = async (req, res) => {
    try {
        let users =[];
        req.body.forEach((user)=> {
            let passwordDigit = md5(user.password);
            users.push([ user.id, user.first_name, user.last_name, user.email, user.gender, user.phone, passwordDigit ]);
        }); 
        // const inputData = [[[id, first_name, last_name, email, gender, phone, passwordDigit]]];
        query = `insert into employee(id, first_name, last_name, email, gender, phone, password) values ?`;
        result = await executor
            .executeQuery(query, [users])
            .catch(function reject(error) {
                throw error;
            });
        // save user token
        return res.status(200).json(responsehandler.makeResponse(result.affectedRows + " rows affected"));

    } catch (error) {
        return res.status(500).json(responsehandler.makeErrorResponse("error while inserting employee data", error.message));
    }
};




const updateEmployee = async (req, res) => {
    try {
        let col = req.body.col;
        let detail = req.body.detail;
        let email = req.email; 
        const inputData = [detail, email];
        query = `update employee set ${col} = ? where email = ?`;
        result = await executor
            .executeQuery(query, inputData)
            .catch(function reject(error) {
                console.log(query);
                throw error;
            });

        return res.status(200).json(responsehandler.makeResponse(result.affectedRows + " rows affected"));
    } catch (error) {
        return res.status(500).json(responsehandler.makeErrorResponse("error while updating employee data", error.message));
    }
};


const deleteEmployee = async (req, res) => {
    try {
        let email = req.email;
        const inputData = [email];
        query = `delete from employee where email = ?`;

        result = await executor
            .executeQuery(query, inputData)
            .catch(function reject(error) {
                throw error;
            });
        return res.status(200).json(responsehandler.makeResponse(result.affectedRows + " rows affected"));
    } catch (error) {
        return res.status(500).json(responsehandler.makeErrorResponse("error while deleting employee data", error.message));
    }
    // return res.status(200).send(result)
    //     }
    // catch(err){
    //     let errorInstance = errorHandle.errorHandle(500,"Internal server error",err)
    //     next(errorInstance)
    // }
};

// deleteEmployee = async function () {
//     try {
//         let coloumn = prompt("Enter the coloumn to delete: ");
//         let detail = prompt("Enter the delete to delete: ");
//         console.log('hi');
//         const inputData = [coloumn, detail];
//         query = `delete from employee where ? = ?`%$[coloumn],$[detail];
//         result = await executor
//             .executeQuery(query, inputData)
//             .catch(function reject(error) {
//                 throw error;
//             });
//         console.log(query);
//         return responsehandler.makeResponse(result.affectedRows + " rows affected");
//     } catch (error) {
//         return responsehandler.makeErrorResponse("error while deleting employee data", error.message);
//     }
// };


const truncateEmployee = async () => {
    try {
        result = await executor
            .executeQuery("truncate table employee")
            .catch(function reject(error) {
                throw error;
            });
        return res.status(200).json(responsehandler.makeResponse("query successful!!", result));
    } catch (error) {
        return res.status(500).json(responsehandler.makeErrorResponse("error while truncating employee data", error.message));
    }
};


const signinEmployee = async (req, res) => {
    try {
        //console.log(req.body)
        let email = req.body.email;
        let password = req.body.password;
        let passwordDigit = md5(password);
        const inputData = [email];
        if (!(email && password)) {
            res.status(400).json(responsehandler.makeErrorResponse("All input is required SOMETHING IS MISSING", error.message));
        }
        else if (email && passwordDigit) {
            query = `SELECT password from employee WHERE email = ?`;
            console.log(query)
            let result = await executor.executeQuery(query, inputData)
                .catch(function reject(error) {
                    throw error;
                });
            console.log(result)
            if (Object.keys(result).length == 0) {
                errorCode = 404;
                throw new Error("username not valid");
            } else {
                if (passwordDigit == result[0].password) {

                    const tokens = jwt.sign(
                        { email },
                        config.token,
                        {
                            expiresIn: "5m",
                        }
                    );
                    return res.status(200).json(responsehandler.makeTokenResponse("signed in sucessfull", tokens));
                }
                else {
                    throw new Error("password not valid");
                }
            }
        }
        else {
            return res.status(500).json(responsehandler.makeErrorResponse("error while signing in  employee data", error.message));
        }
    } catch (error) {
        return res.status(500).json(responsehandler.makeErrorResponse("error while signing in  employee data", error.message));
    }
};




module.exports = {
    readEmployee,
    insertEmployee,
    signupEmployee,
    updateEmployee,
    deleteEmployee,
    truncateEmployee,
    readEmployeeid,
    signinEmployee
};



// jwt authoerization     compl
//  with expire time      compl
//  compression zlib      -----